File: IrrgartenP2/IrrgartenJava/src/irrgartenjava/Dice.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgartenjava;

/**
 *
 * @author moham
 */
// Dice.java
import java.util.Random;

public class Dice {
    private static final int MAX_USES = 5;
    private static final float MAX_INTELLIGENCE = 10.0f;
    private static final float MAX_STRENGTH = 10.0f;
    private static final double RESURRECT_PROB = 0.3;
    private static final int WEAPONS_REWARD = 2;
    private static final int SHIELDS_REWARD = 3;
    private static final int HEALTH_REWARD = 5;
    private static final float MAX_ATTACK = 3.0f;
    private static final float MAX_SHIELD = 2.0f;

    private static final Random generator = new Random();

    public static int randomPos(int max) {
        return generator.nextInt(max);
    }

    public static int whoStarts(int nplayers) {
        return generator.nextInt(nplayers);
    }

    public static float randomIntelligence() {
        return generator.nextFloat() * MAX_INTELLIGENCE;
    }

    public static float randomStrength() {
        return generator.nextFloat() * MAX_STRENGTH;
    }

    public static boolean resurrectPlayer() {
        return generator.nextDouble() < RESURRECT_PROB;
    }

    public static int weaponsReward() {
        return generator.nextInt(WEAPONS_REWARD + 1);
    }

    public static int shieldsReward() {
        return generator.nextInt(SHIELDS_REWARD + 1);
    }

    public static int healthReward() {
        return generator.nextInt(HEALTH_REWARD + 1);
    }

    public static float weaponPower() {
        return generator.nextFloat() * MAX_ATTACK;
    }

    public static float shieldPower() {
        return generator.nextFloat() * MAX_SHIELD;
    }

    public static int usesLeft() {
        return generator.nextInt(MAX_USES + 1);
    }

    public static float intensity(float competence) {
        return generator.nextFloat() * competence;
    }

    public static boolean discardElement(int usesLeft) {
        if (usesLeft == 0) {
            return true;
        }
        if (usesLeft >= MAX_USES) {
            return false;
        }
        return generator.nextDouble() > (double) usesLeft / MAX_USES;
    }
}

File: IrrgartenP4/irrgartenRuby/irrgarten/dice.rb
module Irrgarten
  class Dice
    MAX_USES = 5
    MAX_INTELLIGENCE = 10.0
    MAX_STRENGTH = 10.0
    RESURRECT_PROB = 0.3
    WEAPONS_REWARD = 2
    SHIELDS_REWARD = 3
    HEALTH_REWARD = 5
    MAX_ATTACK = 3.0
    MAX_SHIELD = 2.0

    @generator = Random.new

    class << self
      def random_pos(max)
        @generator.rand(max)
      end

      def who_starts(nplayers)
        @generator.rand(nplayers)
      end

      def random_intelligence
        @generator.rand * MAX_INTELLIGENCE
      end

      def random_strength
        @generator.rand * MAX_STRENGTH
      end

      def resurrect_player
        @generator.rand < RESURRECT_PROB
      end

      def weapons_reward
        @generator.rand(WEAPONS_REWARD + 1)
      end

      def shields_reward
        @generator.rand(SHIELDS_REWARD + 1)
      end

      def health_reward
        @generator.rand(HEALTH_REWARD + 1)
      end

      def weapon_power
        @generator.rand * MAX_ATTACK
      end

      def shield_power
        @generator.rand * MAX_SHIELD
      end

      def uses_left
        @generator.rand(MAX_USES + 1)
      end

      def intensity(competence)
        @generator.rand * competence
      end

      def discard_element(uses_left)
        return true if uses_left == 0
        return false if uses_left >= MAX_USES
        @generator.rand > (uses_left.to_f / MAX_USES)
      end
      
      def preferred_direction(preferred, valid_moves, intelligence)
        if @generator.rand * MAX_INTELLIGENCE < intelligence
          preferred
        else
          valid_moves.sample(random: @generator)
        end
      end
    end
  end
end